# 3. Linear Regression
- Is there a relationship between?
- How strong is the relationship?
- Which features are associated with the target?
- How large the association?
- How accurately can we predict target?
- Is the relationship linear?
- Is there synergy among features?


## 3.1 Simple linear regression
- predicting a quantitative response Y on the basis of a single predictor X. assumes that there is approximately a linear relationship between X and Y.
$$ Y \approx \beta_0 + \beta_1X $$
  - $ \beta_0$ : intercept
  - $ \beta_1$ : slope

### 3.1.1 Estimating the coefficients 
- to obtain coefficient estimates $\hat{\beta}_0$ and $\hat{\beta}_1$ such that the linear model fits the available data well.
  - the most common approach involves minimizing the least squares criterion
    $e_i = y_i - \hat{y}_i$ : residual
    $$ RSS = e^2_1 + e^2_2 + ... + e^2_n = (y_1-\hat{\beta}_0-\hat{\beta}_1x_1)^2 + (y_2-\hat{\beta}_0-\hat{\beta}_1x_2)^2 + ... + (y_n-\hat{\beta}_0-\hat{\beta}_1x_n)^2 \\ \hat{\beta_1} = \frac{\sum^n_{i=1}(x_i-\bar{x})(y_i-\bar{y})}{\sum^n_{i=1}(x_i-\bar{x})^2} \\ \hat{\beta_0}=\bar{y}-\hat{\beta_1}\bar{x}$$

### 3.1.2 Assessing the accuracy of the coefficient estimates
$$ Y = \beta_0 + \beta_1X+\epsilon $$
- $\epsilon$ : catch-all for what we miss with this simple model
- The sample mean and the population mean are different, but in general hte smaple mean will provide a good estimate of the population mean.
- If we use the sample mean $\hat{\mu}$ to estimate $\mu$, this istemate is *unbiased*, in the sense that on average, we expect $\hat{\mu}$ to equal $\mu$
  - On the basis of one particular set of observations, $\hat{\mu}$ might overestimate of underestimate $\mu$, then this average would exactly equal $\mu$
- How accurate is the sample mean $\hat{\mu}$ as an estimate of $\mu$?
  - standard error of $\hat\mu$ $$Var{\hat\mu}=SE(\hat\mu)^2=\frac{\sigma^2}{n} $$
    - the average amount that his estimate $\hat\mu$ differs from the actual $\mu$
  $$ SE(\hat\beta_0)^2 = \sigma^2[\frac{1}{n}+\frac{\bar{x}^2}{\sum^n_{i=1}(x_i-\bar{x})^2}], SE(\hat\beta_1)^2 = \frac{\sigma^2}{\sum^n_{i=1}(x_i-\bar{x})^2} $$
    - SE is smaller when $x_i$ are more spread out; intuitiely we have more leverage to estimate a slope.
  - In general, $\sigma^2$ is not konwn, bur can be estimated from the data $RSE = \sqrt{RSS/(n-2)}
- Standard erros can be used to compute *confidence intervals*
  - 95% confidence interval : a range of values such that with 95% probability, the range will contain the true unknown value of the parameter
    - If we take repeated samples and construct the confidnece interval for each sample, 95% of the invercals will contain the true unknown value of the parameter
    - $\hat\beta_1 \pm 2 \dot SE(\hat\beta_1)$
- Standard errors can also be used to perform hypothesis tests on the coefficients
  - $H_0$ : There is no relationship between X and Y, $\beta_1=0$
  - $H_1$ : There is some relationship between X and Y, $\beta_0\neq 0$
  - need to determine whether $\hat\beta_1$ is sufficiently far from zero that we can confident that $\beta$ is non-zero.
    - depends on the accuracy of $\hat\beta_1$ which depends on SE($\hat\beta_1$) : if $\hat\beta_1$ is small, then even relatively small values of $\hat\beta_1$ may probide strong evidence that $\beta_1 \neq 0$. If SE($\hat\beta_1$) is large, $\hat\beta_1$ must be large in absolute value to reject the null hypothesis
    - use *t-statistic* : $t = \frac{\hat\beta_1-0}{SE(\hat\beta_1)} $=> measures standard deviation that \hat\beta_1 is away from 0.
      - if there is no relationship, a t-distribution with n-2 degrees of freedom
    - t-distribution has a bell shape and for values greather than 30 is similar to the stadard normal distribution.
    - a small p-value indicates that it is unlikely to observe such a substantial association between the predictor and the response due to cahnce, in the absence of any real association between the predictor and the response => there is an association between the predictor and teh response. => **reject null hypothesis**

### 3.1.3 Assessing the accuracy of the model
- Want to quantify the extent to which the model fits the data => residual standard error (RSE) & $R^2$
- Residual Standard Error : estimate of the standard deviation of $\epsilon$
  - actual target deviates from the regression line by RSE units, on average
  - even if the model were correct and the coefficients were known exactly, prediction would still be off by RSE units on average.
  - a measure of the lack of fit of the model to the data.
- $R^2$ statistic : proportion of variance explained
  - a value between 0 and 1 & independent of the scale of Y
    $$ R^2 = \frac{TSS-RSS}{TSS} = 1-\frac{RSS}{TSS} \\ TSS = \sum(y_i - \bar{y})^2,\ total\ sum\ of\ squares$$
  - TSS : amount of variability ingerent in the response before the regression is performed
  - RSS : amount of variability that is left unexplained after performing regression.
- correlation : measure of the linear relationship between X and Y
  $$ Cor(X,Y) = \frac{\sum^n_{i=1}(x_i-\bar{x})(y_i-\bar{y})}{\sqrt{\sum^n_{i=1}(x_i-\bar{x})^2}\sqrt{\sum^n_{i=1}(y_i-\bar{y})^2}}
  - correlation quantifies the association between a single pair of variables rather than between a larger number of variables.

## 3.2 Multiple linear regression
- need to assess the relationship to the response with multiple predictors
- separate simple linear regressions 
  - unclear how to make a single prediction of target given the three predictors since each is associated with a separate regression.
  - each regression ignores other predictors for estimating coefficients.
- extend the simple linear regression so that it can accomodate multiple predictors by giving each predictor a separate slope coefficient in a single model.
  $$ Y = \beta_0 + \beta_1X_1 + \beta_2X_2 + ... + \beta_pX_p+\epsilon

### 3.2.1 Estimating the regression coefficients
- parameters are estimated with the same least squares approach to minimize the sum of squared residuals
  $$ RSS = \sum^n_{i=1}(y_i-\hat{y_i})^2 = \sum^n_{i=1}(y_i-\hat\beta_0-\hat\beta_1x_{i1}-\hat\beta_2x_{i2}-...-\hat\beta_px_{ip})^2
- simple regression coefficient : represent relationship ignoring other predictors
- multiple regression coefficient : represent relationship holding other predictors


### 3.2.2 Some important questions
1. Is at least one of the predictors X1,X2, . . . ,Xp useful in predicting the response?
2. Do all the predictors help to explain Y , or is only a subset of the predictors useful?
3. How well does the model fit the data?
4. Given a set of predictor values, what response value should we predict,and how accurate is our prediction?


1) Is there a relationship between the response and predictors?
- $$ H_0 : \beta_1 = \beta_2=...=\beta_p=0 \\ H_1 : at\ least\ onw\ \beta_j\ is\ nonzero $$
- computing F-statistic $ F=\frac{(TSS-RSS)/p}{RSS/(n-p-1)} $
  - linear model assumption correct : $E[RSS/(n-p-1)]=\sigma^2$
  - provided $H_0$ true : $E[(TSS-RSS)/p]=\sigma^2 $ -> when there is no relationship between the response and predictors, F-statistic values close to 1(분자 분모 모두 분산). / If $H_a$ is true, then $E[(TSS-RSS)/p] > \sigma^2$, F-statistic values larger than 1(분자가 시그마보다 커짐).
- When n is large, an F-statistic just a little larger than 1 still provide evidence against $H_0$. / a larger F-statistic is needed to reject $H_0$ if n is small.
- When $H_0$ is true and the errors $\epsilon_i$ have a normal distribution, F-statistic follows an F-distribution. Compute the p-value associated with the F-statistic.
- WHY DO WE NEED TO LOOK AT THE OVERALL F-STATISTIC when looking p-vals? If any one of p-values for the individual variables is very small, then *at least one of the predictors is related to the response*, but this is flawed, especially when the number of predictors p is large.
  - If p=100 & $H_0$ is true, about 5% of p-values associated with each variable will be below 0.05 by chance. => **we expect to see approximately 5 small p-values even in the absence of any true association**
  - F-statistic does not suffer from this : adjusts for the number of predictors
- If p > n, we cannot fit the multiple linear regression model using least squares, so F-statistic cannot be used.

2) Deciding on important variables
- looking only individual p-values makes false discoveries.
- variable selection : determine which predictors are associated with the response. -> $2^p$ models are needed. efficient approach to choose a smaller set of models to consider.
  - forward selection : begin with null model, then fit p simple linear regressions and add to the null model the variable that results in the lowest RSS
  - backward selection : begin with all variables, then remove the variable with the largest p-value. New (p-1) variable model is fit.
  - mixed selection : begin with no variables, add the variable that provides the best fit. If at any point the p-value for one of the variables in the model rises above a certain threshold, then remove that varaible from the mdoel.

3) Model fit
- $R^2$ increases with added predictors despite no real improvement -> likely lead to poor results on independent test samples due to overfitting
- non-linear pattern suggests a synergy or interaction effect between the predictors.

4) Predictions
- $ f(X) = \beta_0 + \beta_1X_1 + ... + \beta_pX_p $
- Due to irreducible error, we cannot perfectly predict. Prediction intervals are always wider than confidence intervals since they incorporate both the error in the estimate and the uncertainty as to how much individual point will differ from the popultion plane

## 3.3 Other Considerations in the regression model
### 3.3.1 Qualitative predictors
- predictors with only two levels : 1,0 
- predictors with more than two levels : create additional dummy variables

### 3.3.2 Extenstions of the linear model
- linear model assumption : **relationship between the predictors and response are additive and linear**
  - additivity : the association with predictor and response does not dpend on other predictors
  - linearlity : the change in the response associated with a one-unit change in predictor is constant, regardless of the value of X
- removing additive assumption
  - add interaction term by computing product of $X_1$ & $X_2$
  $$ Y = \beta_0 + (\beta_1 + \beta_3X_2)X_1 + \beta_2X_2 + \epsilon \\ = \beta_0 + \bar{\beta_1}X_1+\beta_2X_2+\epsilon
  - \bar{\beta_1}$$ is now a function of $X_2$, the association between $X_1$ and $Y$ is no longer constant
  - **Hierarchical principle : if we include an interaction in a model, we should also include the main effects, even if the p-values associated with their coefficients are not significant.**
- Non-linear Relationships
  - polynomial regression

### 3.3.3 Potential problems
1. Non-linearity of the data
   - If the true relationship is far from lienar, then virtually all of the conclusions that we draw from the fit are suspect
   - Residual plots (residuals vs predicted values): Ideally the residual plot will show no discernible pattern.
2. Correlation of error terms.
   - the error terms are uncorrelated : standard errors are based on the assumption of uncorrelated error terms.
   - such correlation freqeuntly occur in time-series data
     - obtained at adjacent time points will have positively correlated errors
     - plot residuals
3. Non-constant variance of error terms.
- the error terms have a constant variance $Var(\epsilon_i)=\sigma^2$ -> confidence intervals and hypothesis tests rely upon this assumption.
- residual plot -> use concave function transformation log Y or $\sqrt{Y}$
4. Outliers.
- remove them if they have little effect on the least squares line
- studentized residuals : computed by dividing each residual $\epsilon_i$ by its estimated stadrd error (-2 ~ 2) 
5. High-leverage points.
- leverage statistic : $h_i=\frac{1}{n}+\frac{(x_i-\bar{x})^2}{\sum^n_{i'=1}(x_{i'}-\bar{x})^2}$

6. Collinearity.
- two or more predictor variables are closely related to one another
- collinearity reduces the accuracy of the estimated coefficients, their standard error to grow
  - t-statistic is calculated by dividing coefficient by its standard error -> collinearity results in a decline in the t-statistic -> may fail to reject $H_0$
- VIF(variance inflation factor) : the ratio of the variance of coefficient when fitting the full model divided by the variance of coefficient if fit on its on. larger than 5 or 10 indicates collinearity
  $$VIF(\hat{\beta_j}=\frac{1}{1-R^2_{X_j|X_{-j}}}) $$



## 3.6 Lab : Linear Regression

In [2]:
import pandas as pd
import statsmodels.api as sm

df = pd.read_csv("../data/boston.csv", index_col='Unnamed: 0')
df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   lstat  medv  
1   4.98  24.0  
2   9.14  21.6  
3   4.03  34.7  
4   2.94  33.4  
5   5.33  36.2

In [18]:
x = df['lstat']
y = df['medv']

model = sm.OLS.from_formula('medv ~ lstat', data=df)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.544
Model:                            OLS   Adj. R-squared:                  0.543
Method:                 Least Squares   F-statistic:                     601.6
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           5.08e-88
Time:                        14:34:12   Log-Likelihood:                -1641.5
No. Observations:                 506   AIC:                             3287.
Df Residuals:                     504   BIC:                             3295.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     34.5538      0.563     61.415      0.0

In [9]:
### 3.6.3 Multiple linear regression
reg = sm.OLS.from_formula("medv ~ lstat + age", data=df)
result = reg.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.551
Model:                            OLS   Adj. R-squared:                  0.549
Method:                 Least Squares   F-statistic:                     309.0
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           2.98e-88
Time:                        14:11:14   Log-Likelihood:                -1637.5
No. Observations:                 506   AIC:                             3281.
Df Residuals:                     503   BIC:                             3294.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     33.2228      0.731     45.458      0.0

In [14]:
predictors

{'age',
 'chas',
 'crim',
 'dis',
 'indus',
 'lstat',
 'medv',
 'nox',
 'ptratio',
 'rad',
 'rm',
 'tax',
 'zn'}

In [21]:

reg = sm.OLS.from_formula('medv~crim+zn+indus+chas+nox+rm+age+dis+rad+tax+ptratio+lstat', data=df)
result = reg.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.728
Method:                 Least Squares   F-statistic:                     113.5
Date:                Sun, 11 Dec 2022   Prob (F-statistic):          2.23e-133
Time:                        14:35:24   Log-Likelihood:                -1504.9
No. Observations:                 506   AIC:                             3036.
Df Residuals:                     493   BIC:                             3091.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     41.6173      4.936      8.431      0.0

In [27]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

pd.DataFrame({'column': column, 'VIF':variance_inflation_factor(reg.exog, i)}
    for i, column in enumerate(reg.exog_names) if column != "Intercept")


column       VIF
0      crim  1.767486
1        zn  2.298459
2     indus  3.987181
3      chas  1.071168
4       nox  4.369093
5        rm  1.912532
6       age  3.088232
7       dis  3.954037
8       rad  7.445301
9       tax  9.002158
10  ptratio  1.797060
11    lstat  2.870777

In [28]:
### 3.6.4 Interaction Terms
print(sm.OLS.from_formula('medv ~ lstat * age', data=df).fit().summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.556
Model:                            OLS   Adj. R-squared:                  0.553
Method:                 Least Squares   F-statistic:                     209.3
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           4.86e-88
Time:                        14:43:34   Log-Likelihood:                -1635.0
No. Observations:                 506   AIC:                             3278.
Df Residuals:                     502   BIC:                             3295.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     36.0885      1.470     24.553      0.0

In [34]:
print(sm.OLS.from_formula('medv ~ lstat + I(lstat*lstat)', data=df).fit().summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.641
Model:                            OLS   Adj. R-squared:                  0.639
Method:                 Least Squares   F-statistic:                     448.5
Date:                Sun, 11 Dec 2022   Prob (F-statistic):          1.56e-112
Time:                        14:56:13   Log-Likelihood:                -1581.3
No. Observations:                 506   AIC:                             3169.
Df Residuals:                     503   BIC:                             3181.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           42.8620      0.872  

In [44]:
import statsmodels.stats.anova as anova

model1 = sm.OLS.from_formula('medv ~ lstat', data=df).fit()
model2 = sm.OLS.from_formula('medv ~ lstat + I(lstat*lstat)', data=df).fit()

print(sm.stats.anova_lm(model1, model2))

   df_resid           ssr  df_diff     ss_diff           F        Pr(>F)
0     504.0  19472.381418      0.0         NaN         NaN           NaN
1     503.0  15347.243158      1.0  4125.13826  135.199822  7.630116e-28


Hypothesis test comparing the two models
- H0 : two models fit the data equally well
- H1 : the full model is superior => F-stat 135 & p-val is virtually zero -> H0 is rejected

In [48]:
import numpy as np

model3 = sm.OLS.from_formula('medv ~ np.log(lstat)', data=df).fit()
print(model3.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.665
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1000.
Date:                Sun, 11 Dec 2022   Prob (F-statistic):          9.28e-122
Time:                        15:09:09   Log-Likelihood:                -1563.6
No. Observations:                 506   AIC:                             3131.
Df Residuals:                     504   BIC:                             3140.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        52.1248      0.965     54.004

In [51]:
carseat_df = pd.read_csv('../data/Carseats.csv')
carseat_df.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes  
3         14   Yes  Yes  
4         13   Yes   No

In [66]:
predictors = set(carseat_df.columns) - set(['Sales'])
features = [col +' + ' for col in predictors]
features_str = ''.join(features)[:-2]

carseats_model = sm.OLS.from_formula('Sales ~ ' + features_str, data=carseat_df).fit()
print(carseats_model.summary())


                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.873
Model:                            OLS   Adj. R-squared:                  0.870
Method:                 Least Squares   F-statistic:                     243.4
Date:                Sun, 11 Dec 2022   Prob (F-statistic):          1.60e-166
Time:                        15:29:19   Log-Likelihood:                -568.99
No. Observations:                 400   AIC:                             1162.
Df Residuals:                     388   BIC:                             1210.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               5.6606    